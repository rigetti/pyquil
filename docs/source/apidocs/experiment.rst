Experiment
==========

The ``experiment`` module offers a schema and utilities for succinctly expressing commonly
used applications and algorithms in near-term quantum programming. A ``TomographyExperiment``
is intended to be consumed by the ``QuantumComputer.experiment`` method.

**NOTE**: When working with the `experiment` method, the following declared memory labels are
reserved:

 - "preparation_alpha", "preparation_beta", and "preparation_gamma"
 - "measurement_alpha", "measurement_beta", and "measurement_gamma"
 - "symmetrization"
 - "ro"

.. currentmodule:: pyquil.experiment

Schema
------

.. autoclass:: pyquil.experiment.TomographyExperiment

   .. rubric:: Methods

   .. autosummary::
        :toctree: autogen
        :template: autosumm.rst

        ~TomographyExperiment.get_meas_qubits
        ~TomographyExperiment.get_meas_registers
        ~TomographyExperiment.generate_experiment_program
        ~TomographyExperiment.build_setting_memory_map
        ~TomographyExperiment.build_symmetrization_memory_maps
        ~TomographyExperiment.generate_calibration_experiment

.. autoclass:: pyquil.experiment.ExperimentSetting

.. autoclass:: pyquil.experiment.ExperimentResult

.. autoclass:: pyquil.experiment.SymmetrizationLevel

.. autoclass:: pyquil.experiment.CalibrationMethod

Utilities
---------

.. autofunction:: pyquil.experiment.bitstrings_to_expectations
.. autofunction:: pyquil.experiment.correct_experiment_result
.. autofunction:: pyquil.experiment.merge_memory_map_lists
.. autofunction:: pyquil.experiment.ratio_variance
.. autofunction:: pyquil.experiment.read_json
.. autofunction:: pyquil.experiment.to_json
