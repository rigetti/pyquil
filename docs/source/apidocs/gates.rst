Gates and Instructions
======================

A :py:class:`~pyquil.quil.Program` is effectively a list of gates and instructions which
can be created using the function documented in this section::

    >>> p = Program()
    >>> p += H(qubit=0)
    >>> p += RY(angle=pi/3, qubit=1)
    >>> p += CNOT(0, 1)
        ...


.. currentmodule:: pyquil.gates

Native gates for Rigetti QPUs
-----------------------------

Physical quantum processors can enact a subset of all named gates. Luckily,
a small set of gates is universal for quantum computation, so all named gates
can be enacted by suitable combinations of physically realizable gates. Rigetti's
superconducting quantum processors can perform :py:func:`RX` with ``angle=+-pi/2`` or
``angle=+-pi``, :py:func:`RZ` with an arbitrary angle, and :py:func:`CZ` interactions
between neighboring qubits. Rigetti QPUs can natively measure in the computational (Z) basis.

.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    RX
    RZ
    CZ
    MEASURE


All gates and instructions
--------------------------

In general, you will write a quantum program using the full suite of Quil gates and instructions
and use :ref:`the Quil compiler <compiler>` to compile programs into the target instruction
set (ISA). The full list of quantum gates and classical Quil instructions is enumerated here.

.. rubric:: Single-qubit gates
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    I
    X
    Y
    Z
    H
    S
    T
    RX
    RY
    RZ
    PHASE

.. rubric:: Multi-qubit gates
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    CZ
    CNOT
    CCNOT
    CPHASE00
    CPHASE01
    CPHASE10
    CPHASE
    SWAP
    CSWAP
    ISWAP
    PSWAP


.. rubric:: Classical instructions
.. autosummary::
    :toctree: autogen
    :template: autosumm.rst

    WAIT
    RESET
    NOP
    HALT
    MEASURE
    NEG
    NOT
    AND
    OR
    IOR
    XOR
    MOVE
    EXCHANGE
    LOAD
    STORE
    CONVERT
    ADD
    SUB
    MUL
    DIV
    EQ
    LT
    LE
    GT
    GE

.. rubric:: Collections
.. autosummary::

    QUANTUM_GATES
    STANDARD_INSTRUCTIONS

