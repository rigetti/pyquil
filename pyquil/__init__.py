from pyquil._version import pyquil_version
from pyquil.quil import Program
from pyquil.api import list_quantum_computers, get_qc

__version__ = pyquil_version
