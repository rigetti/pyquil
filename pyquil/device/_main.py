##############################################################################
# Copyright 2016-2019 Rigetti Computing
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
##############################################################################
import warnings
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional, Tuple, Union

import networkx as nx
import numpy as np

from pyquil.device._isa import Edge, GateInfo, ISA, MeasureInfo, Qubit, isa_from_graph, isa_to_graph
from pyquil.device._specs import Specs, specs_from_graph
from pyquil.noise import NoiseModel

PERFECT_FIDELITY = 1e0
PERFECT_DURATION = 1 / 100
DEFAULT_CZ_DURATION = 200
DEFAULT_CZ_FIDELITY = 0.89
DEFAULT_ISWAP_DURATION = 200
DEFAULT_ISWAP_FIDELITY = 0.90
DEFAULT_CPHASE_DURATION = 200
DEFAULT_CPHASE_FIDELITY = 0.85
DEFAULT_XY_DURATION = 200
DEFAULT_XY_FIDELITY = 0.86
DEFAULT_RX_DURATION = 50
DEFAULT_RX_FIDELITY = 0.95
DEFAULT_MEASURE_FIDELITY = 0.90
DEFAULT_MEASURE_DURATION = 2000


class AbstractDevice(ABC):
    @abstractmethod
    def qubits(self) -> List[int]:
        """
        A sorted list of qubits in the device topology.
        """

    @abstractmethod
    def qubit_topology(self) -> nx.Graph:
        """
        The connectivity of qubits in this device given as a NetworkX graph.
        """

    @abstractmethod
    def get_isa(self, oneq_type: str = "Xhalves", twoq_type: str = "CZ") -> ISA:
        """
        Construct an ISA suitable for targeting by compilation.

        This will raise an exception if the requested ISA is not supported by the device.

        :param oneq_type: The family of one-qubit gates to target
        :param twoq_type: The family of two-qubit gates to target
        """

    @abstractmethod
    def get_specs(self) -> Optional[Specs]:
        """
        Construct a Specs object required by compilation
        """


class Device(AbstractDevice):
    """
    A device (quantum chip) that can accept programs.

    Only devices that are online will actively be
    accepting new programs. In addition to the ``self._raw`` attribute, two other attributes are
    optionally constructed from the entries in ``self._raw`` -- ``isa`` and ``noise_model`` -- which
    should conform to the dictionary format required by the ``.from_dict()`` methods for ``ISA``
    and ``NoiseModel``, respectively.

    :ivar dict _raw: Raw JSON response from the server with additional information about the device.
    :ivar ISA isa: The instruction set architecture (ISA) for the device.
    :ivar NoiseModel noise_model: The noise model for the device.
    """

    def __init__(self, name: str, raw: Dict[str, Any]):
        """
        :param name: name of the device
        :param raw: raw JSON response from the server with additional information about this device.
        """
        self.name = name
        self._raw = raw

        # TODO: Introduce distinction between supported ISAs and target ISA
        self._isa = ISA.from_dict(raw["isa"]) if "isa" in raw and raw["isa"] != {} else None
        self.specs = Specs.from_dict(raw["specs"]) if raw.get("specs") else None
        self.noise_model = (
            NoiseModel.from_dict(raw["noise_model"]) if raw.get("noise_model") else None
        )

    @property
    def isa(self) -> Optional[ISA]:
        warnings.warn("Accessing the static ISA is deprecated. Use `get_isa`", DeprecationWarning)
        return self._isa

    def qubits(self) -> List[int]:
        assert self._isa is not None
        return sorted(q.id for q in self._isa.qubits if not q.dead)

    def qubit_topology(self) -> nx.Graph:
        """
        The connectivity of qubits in this device given as a NetworkX graph.
        """
        assert self._isa is not None
        return isa_to_graph(self._isa)

    def get_specs(self) -> Optional[Specs]:
        return self.specs

    def get_isa(self, oneq_type: Optional[str] = None, twoq_type: Optional[str] = None) -> ISA:
        """
        Construct an ISA suitable for targeting by compilation.

        This will raise an exception if the requested ISA is not supported by the device.
        """
        if oneq_type is not None or twoq_type is not None:
            raise ValueError(
                "oneq_type and twoq_type are both fatally deprecated. If you want to "
                "make an ISA with custom gate types, you'll have to do it by hand."
            )

        def safely_get(attr: str, index: Union[int, Tuple[int, ...]], default: Any) -> Any:
            if self.specs is None:
                return default

            getter = getattr(self.specs, attr, None)
            if getter is None:
                return default

            array = getter()
            if (isinstance(index, int) and index < len(array)) or index in array:
                return array[index]
            else:
                return default

        def qubit_type_to_gates(q: Qubit) -> List[Union[GateInfo, MeasureInfo]]:
            gates: List[Union[GateInfo, MeasureInfo]] = [
                MeasureInfo(
                    operator="MEASURE",
                    qubit=q.id,
                    target="_",
                    fidelity=safely_get("fROs", q.id, DEFAULT_MEASURE_FIDELITY),
                    duration=DEFAULT_MEASURE_DURATION,
                ),
                MeasureInfo(
                    operator="MEASURE",
                    qubit=q.id,
                    target=None,
                    fidelity=safely_get("fROs", q.id, DEFAULT_MEASURE_FIDELITY),
                    duration=DEFAULT_MEASURE_DURATION,
                ),
            ]
            if q.type is None or "Xhalves" in q.type:
                gates += [
                    GateInfo(
                        operator="RZ",
                        parameters=["_"],
                        arguments=[q.id],
                        duration=PERFECT_DURATION,
                        fidelity=PERFECT_FIDELITY,
                    ),
                    GateInfo(
                        operator="RX",
                        parameters=[0.0],
                        arguments=[q.id],
                        duration=DEFAULT_RX_DURATION,
                        fidelity=PERFECT_FIDELITY,
                    ),
                ]
                gates += [
                    GateInfo(
                        operator="RX",
                        parameters=[param],
                        arguments=[q.id],
                        duration=DEFAULT_RX_DURATION,
                        fidelity=safely_get("f1QRBs", q.id, DEFAULT_RX_FIDELITY),
                    )
                    for param in [np.pi, -np.pi, np.pi / 2, -np.pi / 2]
                ]
            if q.type is not None and "WILDCARD" in q.type:
                gates += [
                    GateInfo(
                        operator="_",
                        parameters="_",
                        arguments=[q.id],
                        duration=PERFECT_DURATION,
                        fidelity=PERFECT_FIDELITY,
                    )
                ]
            return gates

        def edge_type_to_gates(e: Edge) -> List[GateInfo]:
            gates: List[GateInfo] = []
            if (
                e is None
                or isinstance(e.type, str)
                and "CZ" == e.type
                or isinstance(e.type, list)
                and "CZ" in e.type
            ):
                gates += [
                    GateInfo(
                        operator="CZ",
                        parameters=[],
                        arguments=["_", "_"],
                        duration=DEFAULT_CZ_DURATION,
                        fidelity=safely_get("fCZs", tuple(e.targets), DEFAULT_CZ_FIDELITY),
                    )
                ]
            if (
                e is None
                or isinstance(e.type, str)
                and "ISWAP" == e.type
                or isinstance(e.type, list)
                and "ISWAP" in e.type
            ):
                gates += [
                    GateInfo(
                        operator="ISWAP",
                        parameters=[],
                        arguments=["_", "_"],
                        duration=DEFAULT_ISWAP_DURATION,
                        fidelity=safely_get("fISWAPs", tuple(e.targets), DEFAULT_ISWAP_FIDELITY),
                    )
                ]
            if (
                e is None
                or isinstance(e.type, str)
                and "CPHASE" == e.type
                or isinstance(e.type, list)
                and "CPHASE" in e.type
            ):
                gates += [
                    GateInfo(
                        operator="CPHASE",
                        parameters=["theta"],
                        arguments=["_", "_"],
                        duration=DEFAULT_CPHASE_DURATION,
                        fidelity=safely_get("fCPHASEs", tuple(e.targets), DEFAULT_CPHASE_FIDELITY),
                    )
                ]
            if (
                e is None
                or isinstance(e.type, str)
                and "XY" == e.type
                or isinstance(e.type, list)
                and "XY" in e.type
            ):
                gates += [
                    GateInfo(
                        operator="XY",
                        parameters=["theta"],
                        arguments=["_", "_"],
                        duration=DEFAULT_XY_DURATION,
                        fidelity=safely_get("fXYs", tuple(e.targets), DEFAULT_XY_FIDELITY),
                    )
                ]
            if (
                e is None
                or isinstance(e.type, str)
                and "WILDCARD" == e.type
                or isinstance(e.type, list)
                and "WILDCARD" in e.type
            ):
                gates += [
                    GateInfo(
                        operator="_",
                        parameters="_",
                        arguments=["_", "_"],
                        duration=PERFECT_DURATION,
                        fidelity=PERFECT_FIDELITY,
                    )
                ]
            return gates

        assert self._isa is not None
        qubits = [
            Qubit(id=q.id, type=None, dead=q.dead, gates=qubit_type_to_gates(q))
            for q in self._isa.qubits
        ]
        edges = [
            Edge(targets=e.targets, type=None, dead=e.dead, gates=edge_type_to_gates(e))
            for e in self._isa.edges
        ]
        return ISA(qubits, edges)

    def __str__(self) -> str:
        return "<Device {}>".format(self.name)

    def __repr__(self) -> str:
        return str(self)


class NxDevice(AbstractDevice):
    """A shim over the AbstractDevice API backed by a NetworkX graph.

    A ``Device`` holds information about the physical device.
    Specifically, you might want to know about connectivity, available gates, performance specs,
    and more. This class implements the AbstractDevice API for devices not available via
    ``get_devices()``. Instead, the user is responsible for constructing a NetworkX
    graph which represents a chip topology.
    """

    def __init__(self, topology: nx.Graph) -> None:
        self.topology = topology

    def qubit_topology(self) -> nx.Graph:
        return self.topology

    def get_isa(
        self, oneq_type: str = "Xhalves", twoq_type: Optional[Union[str, List[str]]] = None
    ) -> ISA:
        return isa_from_graph(self.topology, oneq_type=oneq_type, twoq_type=twoq_type)

    def get_specs(self) -> Specs:
        return specs_from_graph(self.topology)

    def qubits(self) -> List[int]:
        return sorted(self.topology.nodes)

    def edges(self) -> List[Tuple[Any, ...]]:
        return sorted(tuple(sorted(pair)) for pair in self.topology.edges)
