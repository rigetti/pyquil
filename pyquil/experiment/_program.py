##############################################################################
# Copyright 2016-2019 Rigetti Computing
#
#    Licensed under the Apache License, Version 2.0 (the "License");
#    you may not use this file except in compliance with the License.
#    You may obtain a copy of the License at
#
#        http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS,
#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#    See the License for the specific language governing permissions and
#    limitations under the License.
##############################################################################

from typing import Sequence

import numpy as np

from pyquil.gates import MEASURE, RX, RZ
from pyquil.quil import Program


def parameterized_euler_rotations(
    qubits: Sequence[int],
    *,
    prefix: str,
    suffix_alpha: str = "alpha",
    suffix_beta: str = "beta",
    suffix_gamma: str = "gamma",
) -> Program:
    """
    Given a number of qubits (n), build a ``Program`` containing a ZXZXZ-decomposed gate on each
    qubit, where each ``RZ`` is parameterized by declared values with labels given by the "prefix"
    and "suffix" arguments. Put more plainly, the resulting Quil program on n qubits is::

        RZ(alpha_label[0]) 0
        RX(pi/2) 0
        RZ(beta_label[0]) 0
        RX(-pi/2) 0
        RZ(gamma_label[0]) 0
        ...
        RZ(alpha_label[n-1]) n-1
        RX(pi/2) n-1
        RZ(beta_label[0]) n-1
        RX(-pi/2) n-1
        RZ(gamma_label[n-1]) n-1

    :param qubits: The number of qubits (n).
    :param prefix: The prefix for the declared memory region labels. For example, if the prefix
        is "preparation" and the alpha, beta, and gamma suffixes are left as default, the labels
        would be "preparation_alpha", "preparation_beta", and "preparation_gamma".
    :param suffix_alpha: The suffix for the "alpha" memory region label, which corresponds to the
        first (rightmost) ``Z`` in the ZXZXZ decomposition. Defaults to "alpha".
    :param suffix_beta: The suffix for the "beta" memory region label, which corresponds to the
        second (middle) ``Z`` in the ZXZXZ decomposition. Defaults to "beta".
    :param suffix_gamma: The suffix for the "gamma" memory region label, which corresponds to the
        last (leftmost) ``Z`` in the ZXZXZ decomposition. Defaults to "gamma".
    :return: A ``Program`` containing a 3 parameterized ``RZ``s and 2 fixed ``RX``s per qubit.
    """
    alpha_label = f"{prefix}_{suffix_alpha}"
    beta_label = f"{prefix}_{suffix_beta}"
    gamma_label = f"{prefix}_{suffix_gamma}"

    p = Program()

    alpha = p.declare(alpha_label, "REAL", len(qubits))
    beta = p.declare(beta_label, "REAL", len(qubits))
    gamma = p.declare(gamma_label, "REAL", len(qubits))

    for idx, q in enumerate(qubits):
        p += RZ(alpha[idx], q)
        p += RX(np.pi / 2, q)
        p += RZ(beta[idx], q)
        p += RX(-np.pi / 2, q)
        p += RZ(gamma[idx], q)

    return p


def parameterized_single_qubit_state_preparation(
    qubits: Sequence[int], label: str = "preparation"
) -> Program:
    """
    Given a number of qubits, produce a program as in ``parameterized_euler_rotations`` where each
    memory region is prefixed by ``label``, where label defaults to "preparation".

    :param qubits: The number of qubits (n).
    :param label: The prefix to use when declaring memory in ``parameterized_euler_rotations``.
    :return: A parameterized ``Program`` that can be used to prepare a product state.
    """
    return parameterized_euler_rotations(qubits, prefix=label)


def parameterized_single_qubit_measurement_basis(
    qubits: Sequence[int], label: str = "measurement"
) -> Program:
    """
    Given a number of qubits, produce a program as in ``parameterized_euler_rotations`` where each
    memory region is prefixed by ``label``, where label defaults to "measurement".

    :param qubits: The number of qubits (n).
    :param label: The prefix to use when declaring memory in ``parameterized_euler_rotations``.
    :return: A parameterized ``Program`` that can be used to measure in a multi-qubit Pauli basis.
    """
    return parameterized_euler_rotations(qubits, prefix=label)


def parameterized_readout_symmetrization(
    qubits: Sequence[int], label: str = "symmetrization"
) -> Program:
    """
    Given a number of qubits (n), produce a ``Program`` with an ``RX`` instruction on qubits
    0 through n-1, parameterized by memory regions label[0] through label[n-1], where "label"
    defaults to "symmetrization".

    :param qubits: The number of qubits (n).
    :param label: The name of the declared memory region.
    :return: A ``Program`` with parameterized ``RX`` gates on n qubits.
    """
    p = Program()
    symmetrization = p.declare(f"{label}", "REAL", len(qubits))
    for idx, q in enumerate(qubits):
        p += RX(symmetrization[idx], q)
    return p


def measure_qubits(qubits: Sequence[int]) -> Program:
    """
    Given a number of qubits (n), produce a ``Program`` with a ``MEASURE`` instruction on qubits
    0 through n-1, with corresponding readout registers ro[0] through ro[n-1].

    :param qubits: The number of qubits (n).
    :return: A ``Program`` that measures n qubits.
    """
    p = Program()
    ro = p.declare("ro", "BIT", len(qubits))
    for idx, q in enumerate(qubits):
        p += MEASURE(q, ro[idx])
    return p
