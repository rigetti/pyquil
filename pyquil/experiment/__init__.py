__all__ = [
    "bitstrings_to_expectations",
    "CalibrationMethod",
    "correct_experiment_result",
    "Experiment",
    "ExperimentResult",
    "ExperimentSetting",
    "get_results_by_qubit_groups",
    "group_settings",
    "merge_disjoint_experiments",
    "merge_memory_map_lists",
    "minusX",
    "minusY",
    "minusZ",
    "OperatorEncoder",
    "plusX",
    "plusY",
    "plusZ",
    "ratio_variance",
    "read_json",
    "SIC0",
    "SIC1",
    "SIC2",
    "SIC3",
    "SymmetrizationLevel",
    "TensorProductState",
    "to_json",
    "zeros_state",
]

from pyquil.experiment._calibration import CalibrationMethod
from pyquil.experiment._group import (
    merge_disjoint_experiments,
    get_results_by_qubit_groups,
    group_settings,
)
from pyquil.experiment._main import (
    Experiment,
    OperatorEncoder,
    read_json,
    to_json,
)
from pyquil.experiment._memory import merge_memory_map_lists
from pyquil.experiment._result import (
    ExperimentResult,
    bitstrings_to_expectations,
    correct_experiment_result,
    ratio_variance,
)
from pyquil.experiment._setting import (
    _OneQState,
    ExperimentSetting,
    SIC0,
    SIC1,
    SIC2,
    SIC3,
    TensorProductState,
    minusX,
    minusY,
    minusZ,
    plusX,
    plusY,
    plusZ,
    zeros_state,
)
from pyquil.experiment._symmetrization import SymmetrizationLevel
