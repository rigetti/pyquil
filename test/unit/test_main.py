import numpy as np

from pyquil import Program
from pyquil.experiment._main import _remove_reset_from_program
from pyquil.experiment._program import (
    parameterized_readout_symmetrization,
    parameterized_single_qubit_measurement_basis,
    parameterized_single_qubit_state_preparation,
)
from pyquil.experiment import (
    ExperimentSetting,
    TensorProductState,
    Experiment,
    plusZ,
    read_json,
    to_json,
)
from pyquil.gates import MEASURE, RESET, X, Y
from pyquil.paulis import sI, sX, sY, sZ


EXPERIMENT_REPR = """
shots: 1
active reset: disabled
symmetrization: -1 (exhaustive)
calibration: 1 (plus_eigenstate)
program:
   X 0
   Y 1
"""


def test_tomo_experiment():
    expts = [
        ExperimentSetting(TensorProductState(), sX(0) * sY(1)),
        ExperimentSetting(plusZ(0), sZ(0)),
    ]

    suite = Experiment(settings=expts, program=Program(X(0), Y(1)))
    assert len(suite) == 2
    for e1, e2 in zip(expts, suite):
        # experiment suite puts in groups of length 1
        assert len(e2) == 1
        e2 = e2[0]
        assert e1 == e2
    prog_str = str(suite).splitlines()[3:5]
    assert prog_str == EXPERIMENT_REPR.splitlines()[4:6]


def test_tomo_experiment_pre_grouped():
    expts = [
        [
            ExperimentSetting(TensorProductState(), sX(0) * sI(1)),
            ExperimentSetting(TensorProductState(), sI(0) * sX(1)),
        ],
        [
            ExperimentSetting(TensorProductState(), sZ(0) * sI(1)),
            ExperimentSetting(TensorProductState(), sI(0) * sZ(1)),
        ],
    ]

    suite = Experiment(settings=expts, program=Program(X(0), Y(1)))
    assert len(suite) == 2  # number of groups
    for es1, es2 in zip(expts, suite):
        for e1, e2 in zip(es1, es2):
            assert e1 == e2
    prog_str = str(suite).splitlines()[3:5]
    assert prog_str == EXPERIMENT_REPR.splitlines()[4:6]


def test_tomo_experiment_empty():
    suite = Experiment([], program=Program(X(0)))
    assert len(suite) == 0
    assert str(suite.program) == "X 0\n"


def test_experiment_deser(tmpdir):
    expts = [
        [
            ExperimentSetting(TensorProductState(), sX(0) * sI(1)),
            ExperimentSetting(TensorProductState(), sI(0) * sX(1)),
        ],
        [
            ExperimentSetting(TensorProductState(), sZ(0) * sI(1)),
            ExperimentSetting(TensorProductState(), sI(0) * sZ(1)),
        ],
    ]

    suite = Experiment(settings=expts, program=Program(X(0), Y(1)))
    to_json(f"{tmpdir}/suite.json", suite)
    suite2 = read_json(f"{tmpdir}/suite.json")
    assert suite == suite2


DEFGATE_X = """
DEFGATE XGATE:
    0, 1
    1, 0
"""


TRIMMED_PROG = """
DEFGATE XGATE:
    0, 1
    1, 0

X 0
"""


def test_remove_reset_from_program():
    p = Program(DEFGATE_X)
    p += RESET()
    p += X(0)
    new_p = _remove_reset_from_program(p)
    assert "\n" + new_p.out() == TRIMMED_PROG


def test_generate_experiment_program():
    # simplest example
    p = Program()
    s = ExperimentSetting(in_state=sZ(0), out_operator=sZ(0))
    e = Experiment(settings=[s], program=p, symmetrization=0)
    exp = e.generate_experiment_program()
    test_exp = Program()
    ro = test_exp.declare("ro", "BIT")
    test_exp += MEASURE(0, ro[0])
    assert exp.out() == test_exp.out()
    assert exp.num_shots == 1

    # 2Q exhaustive symmetrization
    p = Program()
    s = ExperimentSetting(in_state=sZ(0) * sZ(1), out_operator=sZ(0) * sZ(1))
    e = Experiment(settings=[s], program=p)
    exp = e.generate_experiment_program()
    test_exp = Program()
    test_exp += parameterized_readout_symmetrization([0, 1])
    ro = test_exp.declare("ro", "BIT", 2)
    test_exp += MEASURE(0, ro[0])
    test_exp += MEASURE(1, ro[1])
    assert exp.out() == test_exp.out()
    assert exp.num_shots == 1

    # add shots
    p = Program()
    p.wrap_in_numshots_loop(1000)
    s = ExperimentSetting(in_state=sZ(0), out_operator=sZ(0))
    e = Experiment(settings=[s], program=p, symmetrization=0)
    exp = e.generate_experiment_program()
    test_exp = Program()
    ro = test_exp.declare("ro", "BIT")
    test_exp += MEASURE(0, ro[0])
    assert exp.out() == test_exp.out()
    assert exp.num_shots == 1000

    # active reset
    p = Program()
    p += RESET()
    s = ExperimentSetting(in_state=sZ(0), out_operator=sZ(0))
    e = Experiment(settings=[s], program=p, symmetrization=0)
    exp = e.generate_experiment_program()
    test_exp = Program()
    test_exp += RESET()
    ro = test_exp.declare("ro", "BIT")
    test_exp += MEASURE(0, ro[0])
    assert exp.out() == test_exp.out()
    assert exp.num_shots == 1

    # state preparation and measurement
    p = Program()
    s = ExperimentSetting(in_state=sY(0), out_operator=sX(0))
    e = Experiment(settings=[s], program=p, symmetrization=0)
    exp = e.generate_experiment_program()
    test_exp = Program()
    test_exp += parameterized_single_qubit_state_preparation([0])
    test_exp += parameterized_single_qubit_measurement_basis([0])
    ro = test_exp.declare("ro", "BIT")
    test_exp += MEASURE(0, ro[0])
    assert exp.out() == test_exp.out()
    assert exp.num_shots == 1

    # multi-qubit state preparation and measurement
    p = Program()
    s = ExperimentSetting(in_state=sZ(0) * sY(1), out_operator=sZ(0) * sX(1))
    e = Experiment(settings=[s], program=p, symmetrization=0)
    exp = e.generate_experiment_program()
    test_exp = Program()
    test_exp += parameterized_single_qubit_state_preparation([0, 1])
    test_exp += parameterized_single_qubit_measurement_basis([0, 1])
    ro = test_exp.declare("ro", "BIT", 2)
    test_exp += MEASURE(0, ro[0])
    test_exp += MEASURE(1, ro[1])
    assert exp.out() == test_exp.out()
    assert exp.num_shots == 1


def test_build_experiment_setting_memory_map():
    p = Program()
    s = ExperimentSetting(in_state=sX(0), out_operator=sZ(0) * sY(1))
    e = Experiment(settings=[s], program=p)
    memory_map = e.build_setting_memory_map(s)
    assert memory_map == {
        "preparation_alpha": [0.0],
        "preparation_beta": [np.pi / 2],
        "preparation_gamma": [0.0],
        "measurement_alpha": [0.0, np.pi / 2],
        "measurement_beta": [0.0, np.pi / 2],
        "measurement_gamma": [0.0, -np.pi / 2],
    }


def test_build_symmetrization_memory_maps():
    p = Program()
    s = ExperimentSetting(in_state=sZ(0) * sZ(1), out_operator=sZ(0) * sZ(1))
    e = Experiment(settings=[s], program=p)
    memory_maps = [
        {"symmetrization": [0.0, 0.0]},
        {"symmetrization": [0.0, np.pi]},
        {"symmetrization": [np.pi, 0.0]},
        {"symmetrization": [np.pi, np.pi]},
    ]
    assert e.build_symmetrization_memory_maps([0, 1]) == memory_maps
