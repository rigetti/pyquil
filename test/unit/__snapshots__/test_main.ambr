# serializer version: 1
# name: test_remove_reset_from_program
  '''
  DEFGATE XGATE AS MATRIX:
  	0, 1
  	1, 0
  
  X 0
  
  '''
# ---
